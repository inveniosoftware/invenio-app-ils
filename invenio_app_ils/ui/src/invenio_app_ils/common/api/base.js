import axios from 'axios';
import { sessionManager } from '../../authentication/services';

const token = sessionManager.token ? sessionManager.token.value : '';

const apiConfig = {
  baseURL:
    process.env.NODE_ENV === 'production'
      ? '/api'
      : 'https://127.0.0.1:5000/api',
  headers: {
    Authorization: `bearer ${token}`,
  },
  withCredentials: true,
};

const http = axios.create(apiConfig);

export { http, apiConfig };
