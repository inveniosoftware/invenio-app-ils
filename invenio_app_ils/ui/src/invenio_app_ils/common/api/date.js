import { DateTime } from 'luxon';
import _get from 'lodash/get';

export const fromISO = stringDate => {
  return DateTime.fromISO(stringDate);
};

/**
 * Converts datetime JSON string to luxon Datetime object
 * @param date
 * @returns {DateTime}
 */
export const toISO = date => {
  return date.toISO();
};

/**
 *  Serializes date for display in short format
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toShortDateTime = date => {
  return date.toFormat('yyyy-MM-dd HH:mm');
};

/**
 *  Serializes date for display in short format
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toShortDate = date => {
  return date.toFormat('yyyy-MM-dd');
};

/**
 *  Serializes date for UTC date in short format used to
 *  normalise dates for REST API
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toUTCShortDate = date => {
  if (!(typeof date === 'string' || date instanceof DateTime)) {
    throw Error('Wrong date type provided');
  }
  if (typeof date === 'string') {
    date = fromISO(date);
  }
  return date.toUTC().toFormat('yyyy-MM-dd');
};

/**
 *  Serializes date for display in tables
 */
export function dateFormatter({ col, row }) {
  const dateField = _get(row, col.field);
  return dateField ? toShortDate(fromISO(dateField)) : null;
}
