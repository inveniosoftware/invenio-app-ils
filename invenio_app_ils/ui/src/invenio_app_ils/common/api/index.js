export { document } from './documents/document';
export { http } from './base';
export { item } from './items/item';
export { eitem } from './eitems/eitem';
export { loan } from './loans/loan';
export { patron } from './patrons/patron';
export { location } from './locations/location';
export { internalLocation } from './locations/internalLocation';
export { keyword } from './keywords/keyword';
export { series } from './series/series';
export { related } from './related/related';
