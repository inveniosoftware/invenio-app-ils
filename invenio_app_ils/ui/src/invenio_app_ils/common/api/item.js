import { http } from './base';

const itemURL = '/items/';

const get = itemPid => {
  return http.get(`${itemURL}${itemPid}`);
};

const buildItemsQuery = (documentPid, state, extraQuery) => {
  const qsDoc = documentPid ? `document_pid:${documentPid}` : '';
  const qsState = state ? ` AND state:${state}` : '';
  const qsExtra = extraQuery ? `${extraQuery}` : '';
  return `${qsDoc}${qsState}${qsExtra}`;
};

const fetchItems = (documentPid, state, extraQuery) => {
  const qs = buildItemsQuery(documentPid, state, extraQuery);
  return http.get(`${itemURL}?q=${qs}`);
};

const fetchItemsByDocPid = (documentPid, config) => {
  const qs = `document_pid:${documentPid} AND status:${
    config.items.available.status
  }`;
  return http.get(`${itemURL}?q=${qs}`);
};

export const item = {
  buildItemsQuery: buildItemsQuery,
  fetchItems: fetchItems,
  fetchItemsByDocPid: fetchItemsByDocPid,
  get: get,
  url: itemURL,
};
