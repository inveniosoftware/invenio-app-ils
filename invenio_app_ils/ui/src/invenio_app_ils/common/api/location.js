import { http } from './base';

const locationURL = '/locations/';
const internalLocationURL = '/internal-locations/';

const get = (locationPid = '') => {
  return http.get(`${locationURL}${locationPid}`);
};

const buildInternalLocationsQuery = (locationPid, extraQuery) => {
  const qsDoc = locationPid ? `location_pid:${locationPid}` : '';
  const qsExtra = extraQuery ? `${extraQuery}` : '';
  return `${qsDoc} ${qsExtra}`;
};

const fetchInternalLocations = (locationPid = '', extraQuery = '') => {
  const qs = buildInternalLocationsQuery(locationPid, extraQuery);
  return http.get(`${internalLocationURL}?q=${qs}`);
};

export const location = {
  fetchInternalLocations: fetchInternalLocations,
  get: get,
  url: locationURL,
};
