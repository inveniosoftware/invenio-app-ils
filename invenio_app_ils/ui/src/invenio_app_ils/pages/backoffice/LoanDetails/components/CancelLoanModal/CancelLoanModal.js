import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { Button, Header, Icon, Input, Modal, Popup } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty';

export default class CancelLoanModal extends Component {
  state = {
    open: false,
    showPopup: false,
    value: this.props.value,
  };

  hide = () => this.setState({ open: false, showPopup: false, value: '' });
  show = () => this.setState({ open: true, showPopup: false, value: '' });

  updateInputRef = element => {
    this.inputRef = element
      ? ReactDOM.findDOMNode(element).querySelector('input')
      : null;
  };

  cancel = () => {
    const { value } = this.state;
    if (isEmpty(value)) {
      this.setState({ showPopup: true });
    } else {
      this.props.action(value);
      this.hide();
    }
  };

  handleOnChange = (event, { value }) => {
    const newState = { value };
    if (this.state.showPopup && !isEmpty(value)) {
      newState.showPopup = false;
    }
    this.setState(newState);
  };

  render() {
    const { loan } = this.props;
    return (
      <Modal
        basic
        size="small"
        trigger={<Button primary content="cancel" onClick={this.show} />}
        open={this.state.open}
        onClose={this.hide}
      >
        <Header content={`Cancel Loan #${loan.pid}`} />
        <Modal.Content>
          <p>
            You are about to cancel loan #{loan.pid}. Please enter a reason for
            cancelling this loan.
          </p>
          <Input
            focus
            fluid
            placeholder="Enter a cancel reason..."
            onChange={this.handleOnChange}
            ref={this.updateInputRef}
            value={this.state.value}
          />
          <Popup
            context={this.inputRef}
            content="Please specify a reason."
            position="bottom left"
            open={this.state.showPopup}
          />
        </Modal.Content>
        <Modal.Actions>
          <Button basic color="black" inverted onClick={this.hide}>
            Back
          </Button>
          <Button color="red" inverted onClick={this.cancel}>
            <Icon name="remove" /> Cancel Loan
          </Button>
        </Modal.Actions>
      </Modal>
    );
  }
}

CancelLoanModal.propTypes = {
  action: PropTypes.func.isRequired,
  loan: PropTypes.object.isRequired,
  value: PropTypes.string,
};

CancelLoanModal.defaultProps = {
  value: '',
};
