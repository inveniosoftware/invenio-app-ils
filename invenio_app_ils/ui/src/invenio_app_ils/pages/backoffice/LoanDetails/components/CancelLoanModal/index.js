export { default as CancelLoanModal } from './CancelLoanModal';
