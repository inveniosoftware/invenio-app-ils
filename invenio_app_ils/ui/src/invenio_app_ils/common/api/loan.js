import { http } from './base';
import { toISO } from './date';
import { DateTime } from 'luxon';

const loanURL = '/circulation/loans/';

const get = loanPid => {
  return http.get(`${loanURL}${loanPid}`);
};

const postAction = (
  url,
  pid,
  loan,
  transactionUserPid,
  transactionLocationPid
) => {
  const now = DateTime.local();
  const payload = {
    transaction_user_pid: transactionUserPid,
    patron_pid: loan.patron_pid,
    transaction_location_pid: transactionLocationPid,
    transaction_date: toISO(now),
  };

  if ('item_pid' in loan) {
    payload['item_pid'] = loan.item_pid;
  } else if ('document_pid' in loan) {
    payload['document_pid'] = loan.document_pid;
  } else {
    throw new Error(
      `No 'item_pid' or 'document_pid' attached to loan '${pid}'`
    );
  }

  return http.post(url, payload);
};

const buildLoansQuery = (
  documentPid,
  itemPid,
  state,
  patronPid,
  extraQuery
) => {
  const qsDoc = documentPid ? `document_pid:${documentPid}` : '';
  const qsItem = itemPid ? `item_pid:${itemPid}` : '';
  const qsUser = patronPid ? `patron_pid:${patronPid}` : '';
  const qsDocItem =
    qsDoc && qsItem ? `(${qsDoc} OR ${qsItem})` : `${qsDoc}${qsItem}`;
  const qsDocItemUser =
    qsDocItem && qsUser
      ? `(${qsDocItem} AND ${qsUser})`
      : `${qsDocItem}${qsUser}`;
  const qsState = state ? `AND state:${state}` : '';
  const qsExtra = extraQuery ? `${extraQuery}` : '';
  return `${qsDocItemUser}${qsState} ${qsExtra}`;
};

const fetchLoans = (
  documentPid,
  itemPid,
  sortBy,
  sortOrder,
  patronPid,
  state,
  extraQuery
) => {
  const qs = buildLoansQuery(
    documentPid,
    itemPid,
    state,
    patronPid,
    extraQuery
  );
  const sort =
    sortBy === 'transaction_date' ? `transaction_date` : `start_date`;
  const sortByOrder = sortOrder === 'asc' ? `${sort}:asc` : `${sort}:desc`;
  return http.get(`${loanURL}?q=${qs}&sort:${sortByOrder}`);
};

export const loan = {
  url: loanURL,
  get: get,
  postAction: postAction,
  fetchLoans: fetchLoans,
  buildLoansQuery: buildLoansQuery,
};
