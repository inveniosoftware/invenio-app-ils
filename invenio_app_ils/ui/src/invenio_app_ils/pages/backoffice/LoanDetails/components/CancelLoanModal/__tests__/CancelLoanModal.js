import React from 'react';
import { shallow, mount } from 'enzyme';
import CancelLoanModal from '../CancelLoanModal';
import { item } from '../../../../../../common/api';
import { doesNotReject } from 'assert';

jest.mock('../../../../../../common/config');

const loan = {
  pid: '1',
};

describe('CancelLoanModal tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  it('should load the cancel loan modal component', () => {
    component = shallow(<CancelLoanModal action={() => {}} loan={loan} />);
    expect(component).toMatchSnapshot();
  });

  it('should not perform loan action with an empty reason', () => {
    const mockAction = jest.fn();
    component = mount(<CancelLoanModal action={mockAction} loan={loan} />);
    expect(component.state('open')).toEqual(false);
    component.find('button').simulate('click');
    expect(component.state('open')).toEqual(true);
    const cancelButton = component.find('button.red');
    expect(cancelButton).toHaveLength(1);

    cancelButton.simulate('click');
    expect(mockAction).not.toHaveBeenCalled();
  });

  it('should show a popup if trying to cancel without a reason', () => {
    component = mount(<CancelLoanModal action={() => {}} loan={loan} />);
    component.find('button').simulate('click');
    expect(component.state('showPopup')).toEqual(false);
    let popup = component.find('Popup');
    const cancelButton = component.find('button.red');
    expect(cancelButton).toHaveLength(1);
    expect(popup.prop('open')).toEqual(false);

    cancelButton.simulate('click');
    popup = component.find('Popup');
    expect(component.state('showPopup')).toEqual(true);
    expect(popup).toHaveLength(1);
    expect(popup.prop('open')).toEqual(true);
  });

  it('should perform the loan action given a reason', () => {
    const value = 'test';
    const mockAction = jest.fn();
    component = mount(<CancelLoanModal action={mockAction} loan={loan} />);
    component.find('button').simulate('click');
    component.setState({ value: value });
    expect(component.state('showPopup')).toEqual(false);

    const cancelButton = component.find('button.red');
    cancelButton.simulate('click');
    expect(component.state('showPopup')).toEqual(false);
    expect(mockAction).toHaveBeenCalledWith(value);
  });
});
