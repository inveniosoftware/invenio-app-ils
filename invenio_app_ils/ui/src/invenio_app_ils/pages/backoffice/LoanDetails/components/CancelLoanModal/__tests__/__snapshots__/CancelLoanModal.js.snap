// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelLoanModal tests should load the cancel loan modal component 1`] = `
<Modal
  basic={true}
  centered={true}
  closeOnDimmerClick={true}
  closeOnDocumentClick={false}
  dimmer={true}
  eventPool="Modal"
  onClose={[Function]}
  open={false}
  size="small"
  trigger={
    <Button
      as="button"
      content="cancel"
      onClick={[Function]}
      primary={true}
    />
  }
>
  <Header
    content="Cancel Loan #1"
  />
  <ModalContent>
    <p>
      You are about to cancel loan #
      1
      . Please enter a reason for cancelling this loan.
    </p>
    <Input
      fluid={true}
      focus={true}
      onChange={[Function]}
      placeholder="Enter a cancel reason..."
      type="text"
      value=""
    />
    <Popup
      content="Please specify a reason."
      disabled={false}
      eventsEnabled={true}
      offset={0}
      on={
        Array [
          "click",
          "hover",
        ]
      }
      open={false}
      pinned={false}
      position="bottom left"
    />
  </ModalContent>
  <ModalActions>
    <Button
      as="button"
      basic={true}
      color="black"
      inverted={true}
      onClick={[Function]}
    >
      Back
    </Button>
    <Button
      as="button"
      color="red"
      inverted={true}
      onClick={[Function]}
    >
      <Icon
        as="i"
        name="remove"
      />
       Cancel Loan
    </Button>
  </ModalActions>
</Modal>
`;
